Fix signed integer overflow in `mapreduce_impl`: review required

FedericoStra opened the pull request:
**Warning**: this PR requires review.

---

This PR solves #38627 as regards the code in `reducedim.jl`, `reduce.jl` and `reinterpretarray.jl`. It **does not** solve all the issues with `missing.jl`.

In the code for [`_mapreduce`](https://github.com/FedericoStra/julia/blob/mapreduce/base/missing.jl#L275-L297) and [`mapreduce_impl`](https://github.com/FedericoStra/julia/blob/mapreduce/base/missing.jl#L311-L364) in `missing.jl` there are still several instances of the construct
```julia
while i <= ilast
    ...
    i += 1
end
```

This overflows when `ilast == typemax(Int)`. More seriously, inside the `while` loops there is often `@inbounds A[i]`, which is then an out-of-bounds read!

I'm pretty sure
```julia
while i <= ilast
    @inbounds ai = A[i]
    ai === missing || break
    i += 1
end
```
can be replaced by
```julia
for outer i = i:ilast
    @inbounds ai = A[i]
    ai === missing || break
end
```

But there are still other lines to fix, such as `i > ilast && return nothing`, so I want to think about it a bit more to see if there is a cleaner way to rewrite it altogether.

Also, I have to add tests!

FedericoStra committed:
Fix signed integer overflow in `mapreduce_impl`

FedericoStra committed:
Add tests for overflow in `mapreduce_impl`

FedericoStra commented:
I added some tests that exercise `mapreduce_impl` with `OffsetArray`s. I'm not sure how to test the functionality at the end of [`reinterpretarray.jl`](https://github.com/FedericoStra/julia/blob/mapreduce/base/reinterpretarray.jl#L668-L713).

FedericoStra committed:
Fix _mapreduce and mapreduce_impl for SkipMissing

FedericoStra commented:
I fixed [`_mapreduce`](https://github.com/FedericoStra/julia/blob/mapreduce/base/missing.jl#L275-L297) and [`mapreduce_impl`](https://github.com/FedericoStra/julia/blob/mapreduce/base/missing.jl#L311-L367) for `SkipMissing` in `missing.jl`. I didn't want to take too much liberty modifying the code, so I tried to keep it as close as possible to the original, even though it doesn't turn out to be the nicest possible probably. I also added the relative tests in [`test/missing.jl`](https://github.com/FedericoStra/julia/blob/mapreduce/test/missing.jl#L504-L514).

FedericoStra commented:
I still have a concern of another kind about the code for `SkipMissing`, which has nothing to do with `OffsetArray`s and out-of-bounds indices.

Imagine we have
```julia
v = repeat(Union{Int,Missing}[missing], 1024*1024);
v[end-1] = v[end] = 1;
sm = skipmissing(v);
```
and we call
```julia
mapreduce(identity, +, sm)
```

In the code for [`_mapreduce`](https://github.com/FedericoStra/julia/blob/mapreduce/base/missing.jl#L275-L297) we scan the entire array to find the two non-`missing` entries at the end, then [at the end of the function](https://github.com/FedericoStra/julia/blob/mapreduce/base/missing.jl#L296) we call [`mapreduce_impl`](https://github.com/FedericoStra/julia/blob/mapreduce/base/missing.jl#L311-L367), which scans the array a second time from start to finish!
This is measurable with
```julia
julia> @btime Base._mapreduce(identity, +, IndexLinear(), $sm)
  905.620 μs (4 allocations: 64 bytes)
2

julia> @btime Base.mapreduce_impl(identity, +, $sm, 1, length(v))
  528.742 μs (6 allocations: 96 bytes)
Some(2)
```
which shows that `_mapreduce` takes double the time because of this double scan.

nalimilan commented on base/missing.jl:
```diff
@@ -274,63 +274,72 @@ mapreduce(f, op, itr::SkipMissing{<:AbstractArray}) =
-    i > ilast && return mapreduce_empty(f, op, eltype(itr))
+    ai === missing && return mapreduce_empty(f, op, eltype(itr))
     a1::eltype(itr) = ai
+    i == typemax(typeof(i)) && return mapreduce_first(f, op, a1)
```
Could this even be changed to this?
```suggestion
    i == ilast && return mapreduce_first(f, op, a1)
```

FedericoStra replied:
Yes, sure. Everything works under the assumption that `ifirst <= ilast`, so if `i != ilast` then it must be `i < ilast <= typemax(typeof(i))` and there is no overflow in the subsequent `i += 1`.

nalimilan commented on base/missing.jl:
```diff
@@ -274,63 +274,72 @@ mapreduce(f, op, itr::SkipMissing{<:AbstractArray}) =
-    mapreduce_impl(f, op, A, ifirst, ilast, pairwise_blocksize(f, op))
+function mapreduce_impl(f, op, A::SkipMissing, ifirst::Integer, ilast::Integer)
+    blksize = pairwise_blocksize(f, op)
+    if blksize <= 0
+        error("pairwise_blocksize($f, $op) must be positive")
+    else
```
I don't think this needs to be checked, as it's a purely internal function whose two definitions return positive numbers.

FedericoStra replied:
Makes sense, and it's very unlikely that a user wants to add methods of `pairwise_blocksize` to customize the block size for different pairs of `(f, op)`.

nalimilan replied:
Yes.

nalimilan reviewed (commented):
Thanks! Did you experience this bug in real code?

Yes, the `SkipMissing` method does some redundant work after finding the first two non-missing values, but in practical uses it shouldn't really matter. I guess we could optimize this a bit by passing the index just before the first non-missing entry to `mapreduce_impl`.

FedericoStra commented:
> Did you experience this bug in real code?

No, I detected it purely by visual inspection of the source code, while reading the implementation of various internal functions. I noticed instead the [similar bug](https://github.com/JuliaLang/julia/issues/38627#issuecomment-737492758) in `Base.show_delim_array` while experimenting with this issue (I'm still not addressing that one here).

nalimilan commented:
As this part is very tricky regarding optimizations around `missing`, could you run some benchmarks to check that there's no performance regression? Thanks.

FedericoStra commented:
I did some benchmarks with the following code:

```julia
using BenchmarkTools
using DataFrames
using Random

seed!() = Random.seed!(42)

function random_tuple_vector(p::Real, len::Integer)
    seed!()
    [rand() < p ? (randn(), randn(), randn()) : missing for _ in 1:len]
end

function random_vector(p::Real, len::Integer)
    seed!()
    [rand() < p ? randn() : missing for _ in 1:len]
end

tuplemuladd(t::Tuple) = muladd(t...)

function benchmark(p::Real, len::Integer)
    v = random_vector(p, len)
    sm = skipmissing(v)
    @benchmark sum($sm) seconds=2
end

function benchmark_tuple(p::Real, len::Integer)
    v = random_tuple_vector(p, len)
    sm = skipmissing(v)
    @benchmark mapreduce(tuplemuladd, +, $sm) seconds=2
end

function run_benchmarks()
    probs = [.15, .5, .85]
    lengths = [3; 10 .^(1:8)]

    df = DataFrame(len=Int[], p=Float64[], tuple=Bool[], trial=BenchmarkTools.Trial[])

    for len in lengths, p in probs
        b = benchmark(p, len)
        nt = (len=len, p=p, tuple=false, trial=b)
        println(nt)
        push!(df, nt)
    end

    for len in lengths, p in probs
        b = benchmark_tuple(p, len)
        nt = (len=len, p=p, tuple=true, trial=b)
        println(nt)
        push!(df, nt)
    end

    transform!(df,
        :trial
            =>
        ByRow(t -> [minimum(t).time, median(t).time, mean(t).time, maximum(t).time])
            =>
        [:min, :median, :mean, :max])

    return df
end
```

I ran the benchmarks on 916762a ("before") and 47bf80f ("after") and called the resulting dataframes `bdf` and `adf` respectively.
The following code plots `adf.median / bdf.median` for each of the 6 combinations of `benchmark`/`benchmark_tutple` and `p = 0.15, 0.5, 0.85`:

```julia
df = select(bdf, 1:3)
df.before = bdf.median
df.after = adf.median

function plot_after_before(gk, sdf, ylims)
    plot(sdf.len, repeat([1], nrow(sdf)); color=:gray, linestyle=:dash)
    plot!(sdf.len, sdf.after ./ sdf.before;
        xscale=:log10, ylims=ylims, color=1, title=NamedTuple(gk), legend=false);
end

plots = [plot_after_before(gk, sdf, extrema(df.after ./ df.before))
    for (gk, sdf) in pairs(groupby(df, [:tuple, :p]))]
plot(plots..., layout=(2,3), titlefontsize=10)
```

Values below `1.0` imply a speedup, values above `1.0` a slowdown. Here is the plot:

![after_before](https://user-images.githubusercontent.com/2906478/104032326-4f458100-51ce-11eb-88be-7d71b3b9ebd2.png)

I don't know how reliable this simple benchmark is, but my conclusion from it is that there is very little performance difference.

StefanKarpinski commented:
Bump. What's needed to make progress here?

nalimilan commented:
My [comment above](https://github.com/JuliaLang/julia/pull/38643/files#r546232587) just needs to be adressed AFAICT.

FedericoStra committed:
Do not check `pairwise_blocksize` in `mapreduce_impl`

FedericoStra commented:
@nalimilan With the last commit I addressed your comment, by removing all checks for `pairwise_blocksize` in `mapreduce_impl`.

I also rebased the branch on the current master so it can be merged (there was a conflict in `test/reduce.jl`).

nalimilan commented on base/missing.jl:
```diff
@@ -281,63 +281,68 @@ mapreduce(f, op, itr::SkipMissing{<:AbstractArray}) =
 
-mapreduce_impl(f, op, A::SkipMissing, ifirst::Integer, ilast::Integer) =
-    mapreduce_impl(f, op, A, ifirst, ilast, pairwise_blocksize(f, op))
+function mapreduce_impl(f, op, A::SkipMissing, ifirst::Integer, ilast::Integer)
+    blksize = pairwise_blocksize(f, op)
+    mapreduce_impl(f, op, A, ifirst, ilast, blksize)
+end
```
I don't see the point of changing this now:
```suggestion
mapreduce_impl(f, op, A::SkipMissing, ifirst::Integer, ilast::Integer) =
    mapreduce_impl(f, op, A, ifirst, ilast, pairwise_blocksize(f, op))
```

nalimilan commented on base/reduce.jl:
```diff
@@ -249,15 +249,17 @@ foldr(op, itr; kw...) = mapfoldr(identity, op, itr; kw...)
 
-mapreduce_impl(f, op, A::AbstractArrayOrBroadcasted, ifirst::Integer, ilast::Integer) =
-    mapreduce_impl(f, op, A, ifirst, ilast, pairwise_blocksize(f, op))
+function mapreduce_impl(f, op, A::AbstractArrayOrBroadcasted, ifirst::Integer, ilast::Integer)
+    blksize = pairwise_blocksize(f, op)
+    mapreduce_impl(f, op, A, ifirst, ilast, blksize)
+end
```
```suggestion
mapreduce_impl(f, op, A::AbstractArrayOrBroadcasted, ifirst::Integer, ilast::Integer) =
    mapreduce_impl(f, op, A, ifirst, ilast, pairwise_blocksize(f, op))
```

nalimilan commented on base/reinterpretarray.jl:
```diff
@@ -702,12 +702,14 @@ end
 
-mapreduce_impl(f::F, op::OP, A::AbstractArrayOrBroadcasted, ifirst::SCartesianIndex2, ilast::SCartesianIndex2) where {F,OP} =
-    mapreduce_impl(f, op, A, ifirst, ilast, pairwise_blocksize(f, op))
+function mapreduce_impl(f::F, op::OP, A::AbstractArrayOrBroadcasted, ifirst::SCartesianIndex2, ilast::SCartesianIndex2) where {F,OP}
+    blksize = pairwise_blocksize(f, op)
+    mapreduce_impl(f, op, A, ifirst, ilast, blksize)
+end
```
```suggestion
mapreduce_impl(f::F, op::OP, A::AbstractArrayOrBroadcasted, ifirst::SCartesianIndex2, ilast::SCartesianIndex2) where {F,OP} =
    mapreduce_impl(f, op, A, ifirst, ilast, pairwise_blocksize(f, op))
```

FedericoStra committed:
Simplify `mapreduce_impl`

Avoid using a local variable `blksize` just to call `pariwise_blocksize`.

Co-authored-by: Milan Bouchet-Valat <nalimilan@club.fr>

nalimilan reviewed (approved):
Thanks!

FedericoStra commented:
> Thanks!

No problem. Sorry I forgot those three `blksize`s and wasted your time. It was my mistake while reverting my changes.
